### PR TITLE
Add table kubernetes_storage_class closes #72

### LINKED ISSUE #72: Add table kubernetes_storage_class
**References**

* https://kubernetes.io/docs/concepts/storage/storage-classes/

### PR DESCRIPTION
**Pull request by ParthaI**
# Integration test result 
<details>
  <summary>Results</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging

SETUP: tests/kubernetes_storage_class []

PRETEST: tests/kubernetes_storage_class
Running terraform

TEST: tests/kubernetes_storage_class
Running terraform

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.create-storageclass will be created
  + resource "null_resource" "create-storageclass" {
      + id = (known after apply)
    }

  # null_resource.delay will be created
  + resource "null_resource" "delay" {
      + id = (known after apply)
    }

  # null_resource.get_storageclass will be created
  + resource "null_resource" "get_storageclass" {
      + id = (known after apply)
    }

Plan: 3 to add, 0 to change, 0 to destroy.
null_resource.create-storageclass: Creating...
null_resource.delay: Creating...
null_resource.delay: Provisioning with 'local-exec'...
null_resource.delay (local-exec): Executing: ["/bin/sh" "-c" "sleep 45"]
null_resource.create-storageclass: Provisioning with 'local-exec'...
null_resource.create-storageclass (local-exec): Executing: ["/bin/sh" "-c" "kubectl apply -f /private/var/folders/v1/slpk6xvx5977gdtb0j_397kc0000gn/T/tests/kubernetes_storage_class/terraform/test/storageclass.yaml"]
null_resource.create-storageclass (local-exec): storageclass.storage.k8s.io/mystorage created
null_resource.create-storageclass: Creation complete after 0s [id=5062732200016868058]
null_resource.delay: Still creating... [10s elapsed]
null_resource.delay: Still creating... [20s elapsed]
null_resource.delay: Still creating... [30s elapsed]
null_resource.delay: Still creating... [40s elapsed]
null_resource.delay: Creation complete after 45s [id=4060684303060120347]
null_resource.get_storageclass: Creating...
null_resource.get_storageclass: Provisioning with 'local-exec'...
null_resource.get_storageclass (local-exec): Executing: ["/bin/sh" "-c" "kubectl get storageclass"]
null_resource.get_storageclass (local-exec): NAME        PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
null_resource.get_storageclass (local-exec): mystorage   kubernetes.io/aws-ebs   Retain          Immediate           true                   46s
null_resource.get_storageclass: Creation complete after 2s [id=5765129079613877437]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.


Running SQL query: test-get-query.sql

[
  {
    "allow_volume_expansion": true,
    "name": "mystorage",
    "reclaim_policy": "Retain",
    "volume_binding_mode": "Immediate"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql

[
  {
    "allow_volume_expansion": true,
    "name": "mystorage",
    "reclaim_policy": "Retain",
    "volume_binding_mode": "Immediate"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql

null
✔ PASSED

POSTTEST: tests/kubernetes_storage_class
Running terraform

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.delete-storageclass will be created
  + resource "null_resource" "delete-storageclass" {
      + id = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
null_resource.delete-storageclass: Creating...
null_resource.delete-storageclass: Provisioning with 'local-exec'...
null_resource.delete-storageclass (local-exec): Executing: ["/bin/sh" "-c" "kubectl delete -f /private/var/folders/v1/slpk6xvx5977gdtb0j_397kc0000gn/T/tests/kubernetes_storage_class/terraform/posttest/storageclass.yaml"]
null_resource.delete-storageclass (local-exec): storageclass.storage.k8s.io "mystorage" deleted
null_resource.delete-storageclass: Creation complete after 0s [id=6998597371333855521]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

TEARDOWN: tests/kubernetes_storage_class



SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
> select
  name,
  provisioner,
  reclaim_policy,
  parameters,
  mount_options
from
  kubernetes_storage_class;
+-----------+-----------------------+----------------+----------------+---------------+
| name      | provisioner           | reclaim_policy | parameters     | mount_options |
+-----------+-----------------------+----------------+----------------+---------------+
| mystorage | kubernetes.io/aws-ebs | Retain         | {"type":"gp2"} | ["debug"]     |
+-----------+-----------------------+----------------+----------------+---------------+

Time: 358ms. Rows fetched: 1. Hydrate calls: 0.
> select
  name,
  allow_volume_expansion,
  provisioner,
  reclaim_policy,
  parameters,
  mount_options
from
  kubernetes_storage_class
where
  allow_volume_expansion;
+-----------+------------------------+-----------------------+----------------+----------------+---------------+
| name      | allow_volume_expansion | provisioner           | reclaim_policy | parameters     | mount_options |
+-----------+------------------------+-----------------------+----------------+----------------+---------------+
| mystorage | true                   | kubernetes.io/aws-ebs | Retain         | {"type":"gp2"} | ["debug"]     |
+-----------+------------------------+-----------------------+----------------+----------------+---------------+

Time: 27ms. Rows fetched: 1. Hydrate calls: 0.
> select
  name,
  allow_volume_expansion,
  provisioner,
  reclaim_policy,
  volume_binding_mode
from
  kubernetes_storage_class
where
  volume_binding_mode = 'Immediate';
+-----------+------------------------+-----------------------+----------------+---------------------+
| name      | allow_volume_expansion | provisioner           | reclaim_policy | volume_binding_mode |
+-----------+------------------------+-----------------------+----------------+---------------------+
| mystorage | true                   | kubernetes.io/aws-ebs | Retain         | Immediate           |
+-----------+------------------------+-----------------------+----------------+---------------------+

Time: 23ms. Rows fetched: 1. Hydrate calls: 0.
```
</details>

### CODE REVIEW

**Comment by ParthaI** [comment]
Golang lint fails because the `listK8sDynamicCRDs` function is not called anywhere. I'm not sure if we need this function.

**Comment by misraved** [comment]
@ParthaI I believe the function `listK8sDynamicCRDs` was created for the dynamic tables which were rolled back since the SDK was not ready to handle the changes. I think we can comment out the code for now and proceed with the new table. @bigdatasourav could you please confirm this approach?

**Comment by bigdatasourav** [comment]
@misraved I have raised a [PR](https://github.com/turbot/steampipe-plugin-kubernetes/pull/100) for the go lint issues. Could you please check and merge it first?

File: docs/tables/kubernetes_storage_class.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  name,
+  allow_volume_expansion,
+  provisioner,
+  reclaim_policy,
+  parameters,
+  mount_options
+from
+  kubernetes_storage_class
+where
+  allow_volume_expansion;
```
Comment (bigdatasourav) [inline docs/tables/kubernetes_storage_class.md:33]:
```suggestion
  not allow_volume_expansion;
```

File: kubernetes/table_kubernetes_storage_class.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	// Create client
+	clientset, err := GetNewClientset(ctx, d)
+	if err != nil {
+		plugin.Logger(ctx).Error("kubernetes_storage_class.getK8sStorageClass", "client_error", err)
+		return nil, err
+	}
+
+	name := d.EqualsQuals["name"].GetStringValue()
+
+	// return name is empty
```
Comment (bigdatasourav) [inline kubernetes/table_kubernetes_storage_class.go:147]:
```suggestion
	// check if name is empty
```

File: kubernetes/table_kubernetes_storage_class.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+		Columns: k8sCommonColumns([]*plugin.Column{
+			{
+				Name:        "provisioner",
+				Type:        proto.ColumnType_STRING,
+				Description: "Provisioner indicates the type of the provisioner.",
+			},
+			{
+				Name:        "reclaim_policy",
+				Type:        proto.ColumnType_STRING,
+				Description: "Dynamically provisioned PersistentVolumes of this storage class are created with this reclaimPolicy. Defaults to Delete.",
```
Comment (bigdatasourav) [inline kubernetes/table_kubernetes_storage_class.go:36]: Descriptions are not in the proper format. Please check.

**Review by bigdatasourav — APPROVED** [review]
LGTM

File: docs/tables/kubernetes_storage_class.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  name,
+  provisioner,
+  reclaim_policy,
+  parameters,
+  mount_options
+from
+  kubernetes_storage_class;
+```
+
+### List storage classes that doesn't allow volume expansion
```
Comment (misraved) [inline docs/tables/kubernetes_storage_class.md:20]:
```suggestion
### List storage classes that don't allow volume expansion
```

File: docs/tables/kubernetes_storage_class.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  parameters,
+  mount_options
+from
+  kubernetes_storage_class
+where
+  not allow_volume_expansion;
+```
+
+### List storage classes with immediate volume binding mode
+```sql
```
Comment (misraved) [inline docs/tables/kubernetes_storage_class.md:37]:
```suggestion
### List storage classes with immediate volume binding mode enabled

```sql
```

File: kubernetes/table_kubernetes_storage_class.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+	"github.com/turbot/steampipe-plugin-sdk/v5/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/v5/plugin"
+	"github.com/turbot/steampipe-plugin-sdk/v5/plugin/transform"
+)
+
+func tableKubernetesStorageClass(ctx context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "kubernetes_storage_class",
+		Description: "A StorageClass provides a way for administrators to describe the classes of storage they offer.",
```
Comment (misraved) [inline kubernetes/table_kubernetes_storage_class.go:18]:
```suggestion
		Description: "Storage class provides a way for administrators to describe the classes of storage they offer.",
```

File: kubernetes/table_kubernetes_storage_class.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+			},
+			{
+				Name:        "reclaim_policy",
+				Type:        proto.ColumnType_STRING,
+				Description: "Dynamically provisioned PersistentVolumes of this storage class are created with this reclaimPolicy. Defaults to Delete.",
+			},
+			{
+				Name:        "allow_volume_expansion",
+				Type:        proto.ColumnType_BOOL,
+				Description: "AllowVolumeExpansion shows whether the storage class allow volume expand.",
```
Comment (misraved) [inline kubernetes/table_kubernetes_storage_class.go:41]:
```suggestion
				Description: "AllowVolumeExpansion shows whether the storage class allows volume expand.",
```

File: kubernetes/table_kubernetes_storage_class.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+	input := metav1.ListOptions{
+		Limit: 500,
+	}
+
+	// Limiting the results
+	limit := d.QueryContext.Limit
+	if d.QueryContext.Limit != nil {
+		if *limit < input.Limit {
+			if *limit < 1 {
+				input.Limit = 1
```
Comment (misraved) [inline kubernetes/table_kubernetes_storage_class.go:95]: Do we need this check, since Steampipe SDK is already validating it?

**Review by misraved — COMMENTED** [review]
Please take a look at the review comments. Thanks!!

**Review by misraved — APPROVED** [review]

**Pull request merged**